File: src-tauri/src/commands/others.rs
//! Commandes des mods « autres » (§7.3) : contenus hors voiture/circuit,
//! avec résolution de conflits de fichiers par priorité.

use super::prelude::*;

/// Liste les mods « autres » avec leurs conflits de fichiers détectés.
#[tauri::command]
pub fn list_other_mods(app: AppHandle, db: State<Db>) -> Result<Vec<crate::others::OtherModCard>, String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    crate::others::list_others(&conn, &cfg).map_err(|e| e.to_string())
}

/// Ouvre le dossier de bibliothèque d'un mod « autre » dans l'explorateur.
/// Même rationale que `open_mod_folder` : le chemin est résolu côté Rust depuis
/// l'overlay, jamais reçu du front, donc le scope ACL du plugin `opener` reste
/// fermé.
#[tauri::command]
pub fn open_other_mod_folder(app: AppHandle, db: State<Db>, id: String) -> Result<(), String> {
    let cfg = crate::config::load(&app);
    let path = {
        let conn = db.0.lock().map_err(|e| e.to_string())?;
        crate::others::folder_path(&conn, &cfg, &id)?
    };
    app.opener()
        .open_path(path.display().to_string(), None::<&str>)
        .map_err(|e| e.to_string())
}

/// Marque/démarque un mod « autre » comme prioritaire (§7.3).
#[tauri::command]
pub fn set_other_priority(db: State<Db>, id: String, priority: bool) -> Result<(), String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    crate::others::set_priority(&conn, &id, priority)
}

/// Active un mod « autre » par junction (§7.3).
#[tauri::command]
pub fn activate_other(app: AppHandle, db: State<Db>, id: String) -> Result<crate::others::ActivateOtherResult, String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    crate::others::activate_other(&conn, &cfg, &id)
}

/// Désactive un mod « autre » (§7.3).
#[tauri::command]
pub fn deactivate_other(db: State<Db>, id: String) -> Result<(), String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    crate::others::deactivate_other(&conn, &id)
}

/// Supprime un mod « autre » : jonctions + fichiers + overlay (§7.3).
#[tauri::command]
pub fn delete_other_mod(app: AppHandle, db: State<Db>, id: String) -> Result<(), String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    crate::others::delete_other(&conn, &cfg, &id)
}

// Cinquième exemplaire du même quintuplet que les mods/apps/packs/sons — voir
// la note dans `commands/addons.rs`.

/// Fichiers annexes d'un « autre mod » (§4.5.2) : notice, images livrées à
/// côté d'un mannequin de pilote nu, par exemple.
#[tauri::command]
pub fn list_other_resources(
    app: AppHandle,
    db: State<Db>,
    id: String,
) -> Result<Vec<crate::resources::ResourceFile>, String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    Ok(crate::resources::list_resources(&crate::others::resources_dir(
        &conn, &cfg, &id,
    )?))
}

/// Ouvre une annexe d'« autre mod » avec l'application par défaut de l'OS.
#[tauri::command]
pub fn open_other_resource(app: AppHandle, db: State<Db>, id: String, rel_path: String) -> Result<(), String> {
    let cfg = crate::config::load(&app);
    let dir = {
        let conn = db.0.lock().map_err(|e| e.to_string())?;
        crate::others::resources_dir(&conn, &cfg, &id)?
    };
    let path = crate::resources::resolve_resource_path(&dir, &rel_path)?;
    app.opener()
        .open_path(path.display().to_string(), None::<&str>)
        .map_err(|e| e.to_string())
}

/// Chemin absolu d'une annexe d'« autre mod », pour l'afficher via `asset://`.
#[tauri::command]
pub fn get_other_resource_path(app: AppHandle, db: State<Db>, id: String, rel_path: String) -> Result<String, String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let dir = crate::others::resources_dir(&conn, &cfg, &id)?;
    Ok(crate::resources::resolve_resource_path(&dir, &rel_path)?
        .display()
        .to_string())
}

/// Contenu brut d'une annexe d'« autre mod », pour la prévisualisation.
#[tauri::command]
pub fn read_other_resource(
    app: AppHandle,
    db: State<Db>,
    id: String,
    rel_path: String,
) -> Result<tauri::ipc::Response, String> {
    let cfg = crate::config::load(&app);
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let dir = crate::others::resources_dir(&conn, &cfg, &id)?;
    Ok(tauri::ipc::Response::new(crate::resources::read_resource(
        &dir, &rel_path,
    )?))
}

File: src/lib/others.ts
// Pont typé vers les commandes « Autres mods » (§7.3).
import { convertFileSrc, invoke } from "@tauri-apps/api/core";
import type { ResourceFile } from "$lib/library";

export interface ConflictInfo {
  other_id: string;
  count: number;
}

/** Sur quoi un mod se greffe et ce qu'il fait (REFONTE§2). Recalculé côté
 * Rust à chaque listage — jamais stocké, sauf la correction de l'utilisateur. */
export interface Attachment {
  kind: "CAR" | "TRACK" | "APP" | "GAME" | "STANDALONE";
  target_id: string | null;
  target_name: string | null;
  /** Force du signal, de la plus sûre à la plus faible. `ARCHIVE` est une
   * conjecture assumée : un pack peut livrer une police pour une voiture qu'il
   * ne touche pas autrement. */
  signal: "USER" | "PATH" | "CONFIG_NAME" | "ARCHIVE" | "NONE";
  nature: "DOCUMENT" | "CONTENT" | "APPEARANCE" | "BEHAVIOUR" | "DEPENDENCY" | "UNRECOGNISED";
}

export interface OtherModRow {
  id: string;
  library_path: string;
  source_archive: string | null;
  imported_at: string;
  is_priority: boolean;
  is_active: boolean;
  junctions: string[];
  /** Nom repris à la main (REFONTE§6.1) : ces mods portent des noms
   * d'archive, c'est-à-dire rien de lisible. `null` tant que rien n'a été
   * saisi — le nom affiché reste alors `id`. */
  display_name_user: string | null;
  /** Note libre (REFONTE§9). */
  notes_user: string | null;
  /** Rattachement corrigé à la main (REFONTE§2.3), `null` tant qu'on n'a rien corrigé. */
  attachment_user: string | null;
  /** Rattachement effectif et nature, déduits ou corrigés. */
  attachment: Attachment;
  conflicts: ConflictInfo[];
  /** Fichiers visant une zone qu'un outil externe synchronise
   * (`extension/config/*​/loaded/`, vao-patches) : Content Manager peut y
   * remplacer la version du mod par la sienne (§4.5.3). */
  externally_managed: number;
  /** Chemins réellement posés dans le jeu, relatifs à la racine d'AC (§11).
   * Vide quand rien n'est posé — mod désactivé, ou copie plus ancienne que ce
   * qui tourne déjà (règle d'or n°5). */
  placed: string[];
  /** Zones du jeu touchées par le mod, dans l'ordre des onglets (§7.3).
   * Plusieurs quand il en touche plusieurs — il apparaît alors sous chacune. */
  categories: string[];
}

export interface ActivateOtherResult {
  junctions: number;
  warnings: string[];
}

export function listOtherMods(): Promise<OtherModRow[]> {
  return invoke<OtherModRow[]>("list_other_mods");
}

/** Corrige le rattachement d'un mod « autre » (REFONTE§2.3). Chaîne vide = revenir à
 * la déduction. */
export function setOtherAttachment(id: string, target: string): Promise<void> {
  return invoke<void>("set_other_attachment", { id, target });
}

export function setOtherPriority(id: string, priority: boolean): Promise<void> {
  return invoke<void>("set_other_priority", { id, priority });
}

export function activateOther(id: string): Promise<ActivateOtherResult> {
  return invoke<ActivateOtherResult>("activate_other", { id });
}

export function deactivateOther(id: string): Promise<void> {
  return invoke<void>("deactivate_other", { id });
}

/** Supprime proprement un mod « autre » (jonctions + fichiers + overlay, §7.3). */
export function deleteOtherMod(id: string): Promise<void> {
  return invoke<void>("delete_other_mod", { id });
}

/** Ouvre le dossier de bibliothèque du mod « autre » dans l'explorateur —
 * résolu et ouvert côté backend, comme `openModFolder`. */
export function openOtherModFolder(id: string): Promise<void> {
  return invoke<void>("open_other_mod_folder", { id });
}

// Cinquième exemplaire du même quintuplet que les mods/apps/packs/sons — voir
// `ResourcesBlock.svelte`.

export function listOtherResources(id: string): Promise<ResourceFile[]> {
  return invoke<ResourceFile[]>("list_other_resources", { id });
}

export function openOtherResource(id: string, relPath: string): Promise<void> {
  return invoke<void>("open_other_resource", { id, relPath });
}

export function otherResourcePath(id: string, relPath: string): Promise<string> {
  return invoke<string>("get_other_resource_path", { id, relPath });
}

export async function otherResourceSrc(id: string, relPath: string): Promise<string> {
  return convertFileSrc(await otherResourcePath(id, relPath));
}

export function readOtherResource(id: string, relPath: string): Promise<ArrayBuffer> {
  return invoke<ArrayBuffer>("read_other_resource", { id, relPath });
}
